packages/arb-token-bridge-ui/src/hooks/useNetworks.ts

```diff
@@ -0,0 +1,255 @@
+    // L3
+    [xaiTestnet.id]: xaiTestnet,
+    [stylusTestnet.id]: stylusTestnet,
+    // E2E
+    [local.id]: local,
+    [arbitrumLocal.id]: arbitrumLocal
+  }[chainId]
+
+  return chain ?? mainnet
+}
+
+function getPartnerChainsQueryParams(chainId: ChainId): ChainId[] {
+  try {
+    const partnerChains = getPartnerChainsForChainId(chainId)
+    return partnerChains.map(chain => chain.id)
+  } catch (e) {
+    return []
+  }
+}
+
+const getProviderForChainCache: {
+  [chainId: number]: StaticJsonRpcProvider
+} = {
+  // start with empty cache
+}
+
+function createProviderWithCache(chainId: ChainId) {
+  const chain = getChainByChainId(chainId)
+  const rpcUrl = rpcURLs[chainId]
+  const provider = new StaticJsonRpcProvider(rpcUrl, chainId)
+  getProviderForChainCache[chain.id] = provider
+  return provider
+}
+
+function getProviderForChainId(chainId: ChainId): StaticJsonRpcProvider {
+  const cachedProvider = getProviderForChainCache[chainId]
+
+  if (typeof cachedProvider !== 'undefined') {
+    return cachedProvider
+  }
+
+  return createProviderWithCache(chainId)
+}
+
+export function sanitizeQueryParams({
+  sourceChainId,
+  destinationChainId
+}: {
+  sourceChainId: ChainId | undefined
+  destinationChainId: ChainId | undefined
+}): {
+  sourceChainId: ChainId
+  destinationChainId: ChainId
+} {
+  // when both `sourceChain` and `destinationChain` are undefined, default to Ethereum and Arbitrum One
+  if (
+    typeof sourceChainId === 'undefined' &&
+    typeof destinationChainId === 'undefined'
+  ) {
+    return {
+      sourceChainId: ChainId.Ethereum,
+      destinationChainId: ChainId.ArbitrumOne
+    }
+  }
+
+  // only `destinationChainId` is defined
+  if (
+    typeof sourceChainId === 'undefined' &&
+    typeof destinationChainId !== 'undefined'
+  ) {
+    const [defaultSourceChainId] =
+      getPartnerChainsQueryParams(destinationChainId)
+    return { sourceChainId: defaultSourceChainId!, destinationChainId }
+  }
+
+  // only `sourceChainId` is defined
+  if (
+    typeof sourceChainId !== 'undefined' &&
+    typeof destinationChainId === 'undefined'
+  ) {
+    const [defaultDestinationChainId] =
+      getPartnerChainsQueryParams(sourceChainId)
+    return {
+      sourceChainId: sourceChainId,
+      destinationChainId: defaultDestinationChainId!
+    }
+  }
+
+  if (
+    !getPartnerChainsQueryParams(sourceChainId!).includes(destinationChainId!)
+  ) {
+    const [defaultDestinationChainId] = getPartnerChainsQueryParams(
+      sourceChainId!
+    )
+    return {
+      sourceChainId: sourceChainId!,
+      destinationChainId: defaultDestinationChainId!
+    }
+  }
+
+  return {
+    sourceChainId: sourceChainId!,
+    destinationChainId: destinationChainId!
+  }
+}
+
+export type UseNetworksState = {
+  sourceChain: Chain
+  sourceChainProvider: StaticJsonRpcProvider
+  destinationChain: Chain
+  destinationChainProvider: StaticJsonRpcProvider
+}
+
+export type UseNetworksSetStateParams =
```

Comment:
I think code would be simpler if we just have both `sourceChain` and `destinationChain` be mandatory when updating state

Reply:
Then the call site can do the necessary checks

Reply:
It could, but then we would have this logic in other place (notably in TransferPanel), where we want to be able to set only one network, and useNetworks set the other network accordingly